content/md/posts/clojure-which-execution-option-to-use.md

```diff
@@ -0,0 +1,217 @@
+
+```clojure
+  :project/run
+  {:main-opts ["-m" "practicalli.sudoku-solver"]}
+
+```
+
+Now the project code can be run using the simple command line form
+
+```bash
+clojure -M:project/run
+```
+
+
+### Chaining aliases together
+
+Alises can be used together by chaining their names on the command line
+
+```bash
+clojure -M:env/dev:env/test:lib/cider:repl/rebel
+```
+
+Using several aliases on the command line will merge their `:extra-paths` and `:extra-deps` values.
+
+The `:main-opts` values are not merged and the value from the last `:main-opts` in the alias chain is used with `clojure.main`
+
+Again, if the command line includes the `-m` flag with a namespace, then that namespace is passed to `clojure.main`, ignoring the values from the aliases.
+
+> Aliases can be chained together to merge their configuration, e.g. `clojure -M:env/dev:repl/rebel` will included the configuration from both `:env/dev` and `:repl/rebel`.  With the `-M` flag, only the `:main-opts` from the last alias in the chain is used.  If `-m namespace/name` is also part of the command, this will be used instead of any values from alias `:main-opts` keys.
+
+
+## clojure.exec -X
+
+`-X` flag provides the flexibility to call any fully qualified function, so Clojure code is no longer tied to `-main`
+
+
+
+Arguments are passed as EDN key value pairs and can therefore be included in any order
+
+> The namesapce of the function to be called is automatically loaded using `requiring-resolve!` (TODO: explanation)
+
+Thus you can now define programs as functions that take a map and execute them via clj:
+
+Call the `status` function from the namespace `practicalli.service` in the current project
+
+```bash
+clojure -X practicalli.service/status
+```
+
+Pass arguments to a `start` function in the `practicalli.service` namespace
+
+```bash
+clojure -X practicalli.service/status :port 8080 :join? false
+```
+
+As the arguments are key/value pairs, it does not matter in which order the pairs are used in the command line.
+
+### Built in clojure.exec functions
+Clojure CLI tools has some built in tools under the special `:deps` alias (not to be confused with the `:deps` configuration in a `deps.edn` file)
+
+* `-X:deps mvn-install` - install a maven jar to the local repository cache
+* `-X:deps git-resolve-tags` - Resolve git coord tags to shas and update deps.edn
+* `-X:deps find-versions` - Find available versions of a library
+* `-X:deps prep` - [prepare source code libraries](https://clojure.org/reference/deps_and_cli#prep) in the dependency tree
+
+> See `clojure --help` for other descriptions and examples (or just read this article further)
+
+
+## Tools -T
+
+Install, run and remove a tool that can be used independently from a Clojure project configuration.
+
+The `-T` flag is based on the `clojure.exec` approach, except it only uses the `:extra-deps` and `:extra-path` from the alias, ignoring the `:deps` and `:path` values from the project `deps.edn` file.
```

Comment:
It's encouraged to use `:deps` and `:paths` in the alias, not `:extra-deps` and `:extra-paths` (although both will work).

Reply:
Ah, this is because it replaces (ignores) the main :deps and :paths in a project.
I am not sure I get the point of including a tool as an alias in a deps.edn file, it seems out of place some how.  Lots to learn about tools...